File: flexinput/src/main/java/com/lytefast/flexinput/utils/FileUtils.java
package com.lytefast.flexinput.utils;

import android.content.ContentResolver;
import android.database.Cursor;
import android.net.Uri;
import android.provider.OpenableColumns;
import android.support.annotation.NonNull;

import com.lytefast.flexinput.model.Attachment;

import java.io.File;


/**
 * @author Sam Shih
 */
public class FileUtils {

  @NonNull
  public static Attachment<File> toAttachment(final File f) {
    return new Attachment<>(f.hashCode(), toUri(f), f.getName(), f);
  }

  public static Uri toUri(final File f) {
    // Use parse due to bug with fresco loader: https://github.com/facebook/fresco/issues/1596
    // FIXME 2017-01: when https://github.com/facebook/fresco/issues/1596 is resolve remove
    // Uri fileUri = Uri.fromFile(f);
    return Uri.parse("file://" + f.getAbsolutePath());
  }

  @NonNull
  public static String getFileNameFromUri(ContentResolver contentResolver, @NonNull final Uri uri)
      throws IllegalArgumentException {
    switch (uri.getScheme()) {
      case ContentResolver.SCHEME_FILE:
        File file = new File(uri.getPath());
        return file.getName();
      case ContentResolver.SCHEME_CONTENT:
        Cursor returnCursor =
            contentResolver.query(uri, null, null, null, null);
        try {
          if (returnCursor != null && returnCursor.moveToFirst()) {
            final int columnIndex = returnCursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
            if (columnIndex >= 0) {
              return returnCursor.getString(columnIndex);
            }
          }
        } finally {
          if (returnCursor != null) {
            returnCursor.close();
          }
        }
        // fallthrough
      default:
        return uri.getLastPathSegment();
    }
  }

  public static String getFileSize(final File file) {
    final long sizeInKb = file.length() / 1024;
    if (sizeInKb < 1024) {
      return sizeInKb + "KB";
    }

    final long sizeInMb = sizeInKb / 1024;
    if (sizeInMb < 1024) {
      return sizeInMb + "MB";
    }

    final long sizeInGb = sizeInMb / 1024;
    return sizeInGb + "GB";
  }
}

File: flexinput/src/main/java/com/lytefast/flexinput/utils/FileUtils.kt
package com.lytefast.flexinput.utils

import android.content.ContentResolver
import android.net.Uri
import android.provider.OpenableColumns

import com.lytefast.flexinput.model.Attachment

import java.io.File


/**
 * @author Sam Shih
 */
object FileUtils {

  @JvmStatic
  fun File.toAttachment(): Attachment<File> =
      Attachment(this.hashCode().toLong(), Uri.fromFile(this), this.name, this)

  @JvmStatic
  @Throws(IllegalArgumentException::class)
  fun Uri.getFileName(contentResolver: ContentResolver): String {
    when (this.scheme) {
      ContentResolver.SCHEME_FILE -> {
        val file = File(this.path)
        return file.name
      }
      ContentResolver.SCHEME_CONTENT -> {
        contentResolver.query(this, null, null, null, null)?.use { cursor ->
          if (cursor.moveToFirst()) {
            val columnIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
            if (columnIndex >= 0) {
              return cursor.getString(columnIndex) ?: this.lastPathSegment
            }
          }
        }
        return this.lastPathSegment
      }
      else -> return this.lastPathSegment
    }
  }

  @JvmStatic
  fun File.getFileSize(): String {
    val sizeInKb = this.length() / 1024
    if (sizeInKb < 1024) {
      return "$sizeInKb KB"
    }

    val sizeInMb = sizeInKb / 1024
    if (sizeInMb < 1024) {
      return "$sizeInMb MB"
    }

    val sizeInGb = sizeInMb / 1024
    return "$sizeInGb GB"
  }
}
